torch/distributed/distributed_c10d.py

```diff
@@ -243,13 +243,27 @@ def __init__(self, backend: Union[str, Backend]):
                 "cuda": backend_val,
             }
         else:
-            # custom backend string in format of "{device_type1}:{backend1},{device_type2}:{backend2}"
-            # TODO
-            pass
-
-        required_devices = ["cpu", "cuda"]
-        for device in required_devices:
-            assert device in self.device_backend_map
+            # make sure the backend string is in the correct format:
+            # "{device_type1}:{backend1},{device_type2}:{backend2}"
+            pattern = r'^[a-zA-Z0-9]+:[a-zA-Z0-9]+(,[a-zA-Z0-9]+:[a-zA-Z0-9]+)*$'
+            # check if the test string matches the pattern
+            if not re.match(pattern, backend):
+                raise ValueError(f"""Invalid backend string argument: {backend}.
+                Custom backend string is an experimental feature where the backend string must be in the format:
+                "<device_type1>:<backend1>,<device_type2>:<backend2>...".""")
+
+            # parse the backend string and populate the device_backend_map
+            for device_backend_pair in backend.lower().split(","):
+                device, backend = device_backend_pair.split(":")
+                if device in self.device_backend_map:
+                    raise ValueError(f"Duplicate device type {device} in backend string: {backend}")
+                self.device_backend_map[device] = Backend(backend)
+
+            # ensure cpu and cuda are specified
+            required_devices = ["cpu", "cuda"]
+            for device in required_devices:
+                if device not in self.device_backend_map:
+                    raise ValueError(f"Device type {device} is not specified in backend string: {backend}")
```

Comment:
I wonder if this check would be too restrictive for non-CUDA, non-CPU use cases?

Reply:
Hmm yeah, I suppose that if user doesn't specify "cpu" or "cuda" then process group won't error out unless explicitly called with a "cpu" or "cuda" tensor, so we don't need this. I can remove this check